FIPS compliant `v3.5.17-dd.3` build

aneesh1 opened the pull request:
## What's changing in `.github/workflows/dd-build.yaml`
- Original `build` job is split up into `build-arm64` and `build-amd64`
  - Doing this b/c we're running with `CGO_ENABLED=1` and I was having issues cross-compiling. The `build-arm64` job runs on an `arm64` machine and `build-amd64` job runs on an `amd64` machine. I use a matrix strategy for both rather than hardcoding environment variables.
- Artifacts uploaded separately under different names `etcd_output_arm64` and `etcd_output_amd64`
- Release job depends on both `build-amd64` and `build-arm64`
- Artifact Download Adjustment
  - We introduced a job `Set artifact name` to set the name of the artifact to be downloaded properly in `releaseassets`.
- Combine checksums and then upload (because we're running two separate jobs on two separate runners)
- Renamed `releaseassets` --> `releaseassetsarm`

## New script `./scripts/build-binary-single-target`
- Very similar to `./scripts/build-binary`
- Builds `etcd` and creates a `.tar.gz` release artifact.
- Takes in variables `TARGET_OS` and `TARGET_ARCH` to build the binary
 
## Modifications to `build.sh`
- New variable:`GO_EXPERIMENT` included in `GO_BUILD_ENV`
- Needed so we can pass `GOEXPERIMENT=boringcrypto`
- I tried doing `export GOEXPERIMENT=boringcrypto`, but it didn't work (see https://github.com/DataDog/etcd/commit/57dc0c7f8f38cd55faeffe24b8ee9ee784c7121c and https://github.com/DataDog/etcd/actions/runs/13862426254/job/38793868704)

## Added `fips.go` to `etcd`, `etcdctl`, `etcdutl` `main` packages

Details about cherry-picked commits:
- Trivy caught some security vulnerabilities in the go standard libraries, upgrading to `1.23.7` fixed this
- `scripts/build-binary-single-target` doesn't run with `GOLDFLAGS` so we can inspect the symbols and ensure that the binary is fips compliant
- Needed https://github.com/DataDog/etcd/pull/10/commits/1bb7e7223b29ceb4dc24333037fd4164ca2cb964 and https://github.com/DataDog/etcd/pull/10/commits/d546bb36af6dd1f7121a2c45c1631aa13dac209c b/c of https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/

ahrtr committed:
Bump actions/upload-artifact to v4.6.0

Signed-off-by: Benjamin Wang <benjamin.ahrtr@gmail.com>

ahrtr committed:
Bump actions/download-artifact to v4.1.8

Signed-off-by: Benjamin Wang <benjamin.ahrtr@gmail.com>

joshjms committed:
bump go to v1.23.6

Signed-off-by: joshjms <joshjms1607@gmail.com>

change go directive to 1.23

Signed-off-by: joshjms <joshjms1607@gmail.com>

ivanvc committed:
Bump go toolchain to 1.23.7

Signed-off-by: Ivan Valdes <ivan@vald.es>

aneesh1 committed:
local fork supports building fips compliant binary

aneesh1 commented:
```
% go tool nm etcd-v3.5.17-dd.2-fips-test-linux-arm64/etcd | grep -i 'crypto/internal/boring/sig.FIPSOnly'
  17b500 T crypto/internal/boring/sig.FIPSOnly.abi0
% go tool nm etcd-v3.5.17-dd.2-fips-test-linux-arm64/etcdctl | grep -i 'crypto/internal/boring/sig.FIPSOnly'
  176eb0 T crypto/internal/boring/sig.FIPSOnly.abi0
% go tool nm etcd-v3.5.17-dd.2-fips-test-linux-arm64/etcdutl | grep -i 'crypto/internal/boring/sig.FIPSOnly'
  42c2c0 T crypto/internal/boring/sig.FIPSOnly.abi0
% go tool nm etcd-v3.5.17-dd.2-fips-test-linux-amd64/etcd | grep -i 'crypto/internal/boring/sig.FIPSOnly'
  5fde80 T crypto/internal/boring/sig.FIPSOnly.abi0
% go tool nm etcd-v3.5.17-dd.2-fips-test-linux-amd64/etcdctl | grep -i 'crypto/internal/boring/sig.FIPSOnly'
  5f9c20 T crypto/internal/boring/sig.FIPSOnly.abi0
% go tool nm etcd-v3.5.17-dd.2-fips-test-linux-amd64/etcdutl | grep -i 'crypto/internal/boring/sig.FIPSOnly'
  8cdcc0 T crypto/internal/boring/sig.FIPSOnly.abi0
```

nyodas commented on scripts/build-release-single-target:
```diff
@@ -0,0 +1,100 @@
+
+source ./scripts/test_lib.sh
+
+VER=${1:-}
```
Why not VERSION ?

aneesh1 replied:
Copied from `scripts/build-binary`. My intention was that this script would look the same as `scripts/build-binary`. The only difference is that this one would build and release one image version.

nyodas replied:
👍

badouralix replied:
If we copied `scripts/build-binary`, this file should be named `scripts/build-binary-single-target`, otherwise we'll confuse it with `script/build-release.sh`

nyodas commented on scripts/build-release-single-target:
```diff
@@ -0,0 +1,100 @@
+  local proj=${2}
+
+  if [ ! -d "${proj}" ]; then
+    run git clone "${REPOSITORY}"
```
Why run ? , couldn't git clone directly work ?

aneesh1 replied:
Copied from `scripts/build-binary`
https://github.com/DataDog/etcd/blob/ebb2b06b1cd6356b1e4ad4b9fa87793aa506d051/scripts/build-binary.sh#L19-L21

Same reasoning as the comment above.

nyodas replied:
🦖 👍

nyodas reviewed (approved):
LGTM

badouralix commented on .github/workflows/dd-build.yaml:
```diff
@@ -8,18 +8,69 @@ on:
+        echo "RELEASE_VERSION=${GITHUB_REF#refs/tags/}" >> $GITHUB_ENV
+        echo "TARGET_OS=$(echo ${{ matrix.platform }} | cut -d'/' -f1)" >> $GITHUB_ENV
+        echo "TARGET_ARCH=$(echo ${{ matrix.platform }} | cut -d'/' -f2)" >> $GITHUB_ENV
+        export GOEXPERIMENT=boringcrypto
```
Do we need this export ?

badouralix replied:
Should we remove it if we don't need it ?

badouralix commented on .github/workflows/release.yaml:
```diff
@@ -50,7 +50,7 @@ jobs:
     steps:
       - name: get-image
-        uses: actions/download-artifact@v3
+        uses: actions/download-artifact@fa0a91b85d4f404e444e00e005971372dc801d16 # v4.1.8
```
Should we also pin it in `dd-build.yaml` ?

badouralix commented on .github/workflows/release.yaml:
```diff
@@ -34,7 +34,7 @@ jobs:
         docker image save -o /tmp/etcd-img.tar gcr.io/etcd-development/etcd
     - name: upload-image
-      uses: actions/upload-artifact@v3
+      uses: actions/upload-artifact@65c4c4a1ddee5b72f698fdd19549f0f0fb45cf08 # v4.6.0
```
Since this file is upstream, do we want to modify it ?

aneesh1 replied:
Yes, taken from upstream commit - https://github.com/etcd-io/etcd/commit/21f048ed0ca7a4b38cf086e3edd7637a3511e775.

badouralix commented on build.sh:
```diff
@@ -20,7 +20,10 @@ CGO_ENABLED="${CGO_ENABLED:-0}"
 # shellcheck disable=SC2206
 GO_LDFLAGS=(${GO_LDFLAGS:-} "-X=${VERSION_SYMBOL}=${GIT_SHA}")
-GO_BUILD_ENV=("CGO_ENABLED=${CGO_ENABLED}" "GO_BUILD_FLAGS=${GO_BUILD_FLAGS:-}" "GOOS=${GOOS}" "GOARCH=${GOARCH}")
+# shellcheck disable=SC2206
+GO_ADDITIONAL_VARS="${GO_ADDITIONAL_VARS:-}"
+IFS=' ' read -r -a GO_ADDITIONAL_VARS_ARR <<< "${GO_ADDITIONAL_VARS}"
+GO_BUILD_ENV=("CGO_ENABLED=${CGO_ENABLED}" "${GO_ADDITIONAL_VARS_ARR[@]}" "GO_BUILD_FLAGS=${GO_BUILD_FLAGS:-}" "GOOS=${GOOS}" "GOARCH=${GOARCH}")
```
Since we introduce the change and we need to maintain it, let's go with simplicity and only set `GOEXPERIMENT`

aneesh1 replied:
I want to try and upstream this portion since I believe it could be useful for others. We won't need to maintain it either. What do you think?

badouralix replied:
Better only upstream `GOEXPERIMENT` and introduce `GO_ADDITIONAL_VARS` later on if we need it
`GO_BUILD_FLAGS` is a similar optional envvar, and I think we should follow this pattern for now
The overhead of the bash read array makes the patch harder to maintain, even upstream, without direct benefits to be worth it right now

badouralix commented on .go-version:
```diff
@@ -1 +1 @@
-1.22.9
+1.23.7
```
Should we bump straight to go 1.24 since it has FIPS 140-3 compliance ?

aneesh1 replied:
Good callout. We also get protections against TLS downgrades - we should upgrade.

Upstream has merged https://github.com/etcd-io/etcd/pull/19440 which bumps go to `1.24`. I don't think this has been backported into the `release-3.5` branch. On my local copy I did:
```
$ git log | grep 8e4de789e5a223ef14d0baf579ec772816a51c49
```

and it returned nothing. Let's wait for the commit to be backported to `release-3.5` and then upgrade.

badouralix commented on .github/workflows/dd-build.yaml:
```diff
@@ -8,18 +8,67 @@ on:
+    strategy:
+      matrix:
+        platform: ["linux/amd64"]
+    runs-on: ubuntu-latest
```
I'm thinking we should pin `ubuntu-24.04` to be consistent with `ubuntu-24.04-arm`

badouralix commented on .github/workflows/dd-build.yaml:
```diff
@@ -8,18 +8,67 @@ on:
 jobs:
-  build:
-    runs-on: ubuntu-latest
+  build-arm64:
```
Nit we should define `build-amd64` before `build-arm64` to respect the alphabetical order and for consistency with `needs: [build-amd64, build-arm64]`

badouralix commented on .github/workflows/dd-build.yaml:
```diff
@@ -8,18 +8,67 @@ on:
     - name: Build etcd
-      run: REPOSITORY=https://github.com/${{ env.GITHUB_REPOSITORY}}.git ./scripts/build-binary ${{ env.RELEASE_VERSION }}
+      run: |
+        REPOSITORY=https://github.com/${{ env.GITHUB_REPOSITORY}}.git \
```
```suggestion
        REPOSITORY=https://github.com/${{ env.GITHUB_REPOSITORY }}.git \
```

badouralix commented on .github/workflows/dd-build.yaml:
```diff
@@ -8,18 +8,67 @@ on:
+        echo "TARGET_ARCH=$(echo ${{ matrix.platform }} | cut -d'/' -f2)" >> $GITHUB_ENV
+    - name: Build etcd
+      run: |
+        REPOSITORY=https://github.com/${{ env.GITHUB_REPOSITORY}}.git \
```
```suggestion
        REPOSITORY=https://github.com/${{ env.GITHUB_REPOSITORY }}.git \
```

badouralix commented on build.sh:
```diff
@@ -16,11 +16,12 @@ GOOS=${GOOS:-$(go env GOOS)}
 # shellcheck disable=SC2206
 GO_LDFLAGS=(${GO_LDFLAGS:-} "-X=${VERSION_SYMBOL}=${GIT_SHA}")
-GO_BUILD_ENV=("CGO_ENABLED=${CGO_ENABLED}" "GO_BUILD_FLAGS=${GO_BUILD_FLAGS:-}" "GOOS=${GOOS}" "GOARCH=${GOARCH}")
+GO_BUILD_ENV=("CGO_ENABLED=${CGO_ENABLED}" "GO_BUILD_FLAGS=${GO_BUILD_FLAGS:-}" "GOOS=${GOOS}" "GOARCH=${GOARCH}" ${GOEXPERIMENT:+"GOEXPERIMENT=${GOEXPERIMENT}"})
```
Any reason not to do `"GOEXPERIMENT=${GOEXPERIMENT:-}"` just like with `GO_BUILD_FLAGS`

aneesh1 replied:
No reason. Just confirmed in https://cs.opensource.google/go/go/+/refs/tags/go1.23.7:src/internal/buildcfg/exp.go;l=106-115 that as long as we don't do `GOEXPERIMENT=none`, it should be alright.

badouralix reviewed (commented):
Given we want to build one arch at a time, I don't think we have the choice but to fork `scripts/build-binary` into `scripts/build-binary-single-target` indeed

badouralix commented on build.sh:
```diff
@@ -65,7 +67,7 @@ etcd_build() {
     cd ./server
     # Static compilation is useful when etcd is run in a container. $GO_BUILD_FLAGS is OK
     # shellcheck disable=SC2086
-    run env "${GO_BUILD_ENV[@]}" go build ${GO_BUILD_FLAGS:-} \
+    run env "${GO_BUILD_ENV[@]}" go build ${GO_BUILD_FLAGS} \
```
Why do we need this change compared to upstream ?

aneesh1 replied:
We don't - I was using cursor for something and it added this and I unintentionally pushed it.

aneesh1 committed:
version: bump up to 3.5.17-dd.3

command: DRY_RUN=false TARGET_VERSION="v3.5.17-dd.3" ./scripts/release_mod.sh update_versions

badouralix commented on server/go.mod:
```diff
@@ -26,12 +26,12 @@ require (
-	go.etcd.io/etcd/raft/v3 v3.5.17-dd.1
+	go.etcd.io/etcd/api/v3 v3.5.17-dd.3
+	go.etcd.io/etcd/client/pkg/v3 v3.5.17-dd.3
+	go.etcd.io/etcd/client/v2 v2.305.17-dd.3
```
Do we have this `v2.305.17-dd.3` version ?

aneesh1 replied:
No - let's remove it.
